File: js/task3/S3_PIPE_02.js
// S3_PIPE_02 — compose()

function compose(...fns) {
  return (input) => {
    let value = input;
    for (let i = fns.length - 1; i >= 0; i--) {
      value = fns[i](value);
    }
    return value;
  };
}

// Tests
const add1 = (n) => n + 1;
const double = (n) => n * 2;

const f = compose(add1, double);

console.log(f(3)); // add1(double(3)) => 7

File: python/task3c/S3_PIPE_02.py
# S3_PIPE_02 — Implement compose(*fns)

def compose(*fns):
    def runner(x):
        value = x
        for fn in reversed(fns):
            value = fn(value)
        return value
    return runner


def main():
    add2 = lambda n: n + 2
    times3 = lambda n: n * 3
    minus5 = lambda n: n - 5

    # pipe: left → right
    def pipe(*fns):
        def r(x):
            for fn in fns:
                x = fn(x)
            return x
        return r

    p = pipe(add2, times3, minus5)
    c = compose(add2, times3, minus5)

    print("pipe result:", p(10))      # ((10+2)*3)-5 = 31
    print("compose result:", c(10))   # add2(times3(minus5(10))) = 17


if __name__ == "__main__":
    main()
